### DISCUSSION TITLE
Using PostgreSQL SKIP LOCKED with sqlalchemy.select

### DISCUSSION THREAD

**Discussion by jramnani** [sqlalchemy/sqlalchemy#10460] [Usage Questions] [answered]
Hello,

I'm relatively new to using SQLAlchemy.  While working with PostgreSQL I noticed a discrepancy between the SQL output generated by `sqlalchemy.select` and `Session.query`.

`Session.query` is generating the SQL I want, but that method is marked as deprecated.  `sqlalchemy.select` is close, but missing a clause I need in the output.  I'm wondering if there is a mistake in how I am using `sqlalchemy.select`?

Given the following schema:

```sql
CREATE TABLE IF NOT EXISTS
  jobs(
    id SERIAL PRIMARY KEY NOT NULL,
    type TEXT NOT NULL,
    status TEXT NOT NULL,
  );
```

And this database model object:

```python
class Job(BaseModel):
    __tablename__ = "jobs"

    id: Mapped[int] = mapped_column(primary_key=True)
    type: Mapped[Text]
    status: Mapped[Text]
```

I would like to write the following query:

```
SELECT id, type, status
  FROM jobs
 WHERE status = 'pending' FOR UPDATE SKIP LOCKED LIMIT 1;
```

When using `Session.query` I get the SQL I want.  Here is that query:

```
with self.session as session:
    query = (
                session.query(Job)
                .filter(Job.status == JobStatus.PENDING.value)
                .with_for_update(skip_locked=True)
                .limit(1)
            )
```

And the resulting SQL:

```sql
SELECT jobs.id AS jobs_id, jobs.type AS jobs_type, jobs.status AS jobs_status
  FROM jobs
 WHERE jobs.status = %(status_1)s
 LIMIT %(param_1)s FOR UPDATE SKIP LOCKED
```

When using `sqlalchemy.select` here is the query I used:

```python
select_query = (
    sqlalchemy.select(Job)
    .where(Job.status == JobStatus.PENDING.value)
    .with_for_update(skip_locked=True)
    .limit(1)
    )
```

And here is the resulting SQL.  Notice the absence of `SKIP LOCKED`.

```sql
SELECT jobs.id, jobs.type, jobs.status
FROM jobs
WHERE jobs.status = :status_1
 LIMIT :param_1 FOR UPDATE
```

Is there something I'm doing wrong with my select query?  I don't know when `Session.query` is going to be removed, so I'd like to port my query if I can.

Environment:

* Python: 3.11.4
* SQLAlchemy: 2.0.21
* PostgreSQL: 15.4

Thanks for listening.

**Comment by zzzeek** [comment, marked as answer]
hi -

SKIP LOCKED is postgresql specific so we assume you are saying "print(statement)" and not seeing PostgreSQL-specific syntax.  See the FAQ at https://docs.sqlalchemy.org/en/20/faq/sqlexpressions.html#stringifying-for-specific-databases 
 
add echo=True to your engine and you will see the statement emitted assuming its PG:

```
 SELECT a.id, a.data 
FROM a 
WHERE a.data = %(data_1)s 
 LIMIT %(param_1)s FOR UPDATE SKIP LOCKED

```

**Reply by jramnani** [reply]
Ah. After adding `echo=True` to my `Engine` I see it now in my test output.  Thank you! 🙇 

I have found `print(statement)` to be a really useful feature. I'll need to get used to doing it the way described in the FAQ you linked to. e.g. `print(select_query.compile(my_engine))`

Thanks again! :)
